docs/enqueuingStrategy.md

```diff
@@ -0,0 +1,18 @@
+# Queueing strategy
+
+The moving target of the gas market presents a unique challenge to Ethereum users and developers. Gas prices estimates can quickly become invalidated as the gas market moves. Transactions, that should have been mined in minutes due to the estimate at the time of broadcast, can take days in practice. 
```

Comment:
Ethereum mining is a bin packing problem. 

Miners have a fixed-capacity bin (the block) and a list of pending boxes (transactions) to put in the bin. When the pending pool of transactions exceeds a block's capacity, then miners can be very selective over which transactions they include in their next block. Of course, they have a financial incentive to pack the block with transactions that maximises their profit. 

So it is up to us, the transaction signer, to pick a fee that will entice the miners to include our transaction in their next few blocks. But there is one important constraint that can be problematic for relayers, but miners must follow: 

> Network rule: All transactions from the same signer must be accepted in order and each transaction nonce must strictly increment by one. 

The motivation for the network rule is to ensure transactions are processed in order and miners cannot easily invalidate previously authorised transactions from the same signer. With this in mind, it can result in the following problems: 

- Bad fee management: If we have a queue of transactions (tx1, tx2, ..., txN) and if tx1 has a low fee (5 gwei), then tx1 can prevent all subsequent transactions getting accepted into a block. Why? Miners may only accept transactions with a fee of 10 gwei or more. _Solution: Order all transactions in the queue by fee._ 

- No out-of-order transactions: If we want to send 100 transactions to the network, then they all must be accepted in order. _Solution: Multiple signing keys and queues of transactions._

- Transaction dropped from network: When the network is fully congested, it is possible that your transaction is dropped from the pending pool and thus it will never be confirmed. _Solution: Periodically republish the transaction with a higher fee._

- Prioritise new transactions: If you have already published 100 transactions to the pending pool, but a new transaction appears with a higher priority, then what can you do? _Solution: Re-order all transactions (race condition issues) or publish using a new signing key._

We cover the problems in [more detail here](https://docs.google.com/presentation/d/1gWrEjJICL23583pqKIsKg9HxIFUo6j_AvexrM1GPHvw/edit?usp=sharing), but clearly running a relayer (especially during congestion) has plenty of pitfalls. We hope any.sender will alleviate the above issues for more developers, but let's discuss a bit more on how we have tried to tackle the problems.

Reply:
I'll add a link to the presentation for more detail.

I had a lot of detail like this before, but took it all out. I wanted to keep this simple, and focus on the 2 core problems that we're solving:
* the gas market moves, invalidating estimates
* queueing multiple transactions on the same key can lead to transactions being held up
and how we solve these problems at a high level.

I also dont think we need to explain some obvious things. For example I think there's no need to explain how miners work, every developer knows they prioritise higher fee transactions. And on mainnet the pending pool always exceeds the block size.